### DISCUSSION TITLE
about : processing and gridding gravity data  with fatiando a terra

### DISCUSSION THREAD

**Discussion by prince-Ha** [fatiando/community#80] [Q&A] [answered]
Hello, I know you are very busy, but I am asking you this question because I need your help. After watching the video on YouTube, I am processing the gravitational data with the fatiando tutorial.

 1 . there are many differences between simple Bouguer and complete Bouguer , so I ask you what the problem is.
![CBA_300dpi](https://user-images.githubusercontent.com/69688086/203570287-87c47625-81e7-4173-a854-29b15361aab9.jpg)

2. run Regional and Residual seperation code, the kernel continue to die.

code : deep_sources.fit((data.easting_m, data.northing_m, data.height_geometric_m), data.gravity_bouguer_mgal)

3. I want to know how to make corrections on the coast and the sea.
<img width="867" alt="스크린샷 2022-11-23 오후 10 46 11" src="https://user-images.githubusercontent.com/69688086/203563870-46aa9e00-1852-4304-8a67-ad09f7757db8.png">

I'm worried if it's rude to ask.

Thank you for reading this long article.

I'll attach the file together

thanks

**Comment by santisoler** [comment, marked as answer]
Hi @prince-Ha! Opening this issue is not rude at all. The idea of allowing users to open issues or discussions in our forum is to encourage the conversation within the community. Your questions and inputs are very welcomed 🙂.

**Answer to point 1**

After reading your notebook I noticed that your free-air anomaly data doesn't include the observation height. This height is as important as the observed gravity values: the observed gravity depends on the densities of the bodies that generate it and also on the distance to those bodies. In cell 28 you add a `height_geometric_m` column to your `data` and you use these values when computing the terrain effect as the observation heights. Those heights are the same as the topographic heights in continental regions and bathymetry depths in oceanic regions. If we assume that the gravity observations where taken through a ground survey, it makes sense to assume that the observation heights will be approximately the topographic height. But I suspect that in the oceanic regions they weren't taken in the deepest point of the sea, but rather at sea level. So the observation heights you choose might be wrong and this might be the source of this difference between the simple Bouguer correction and the terrain effect. I would also recommend you to plot the Bouguer correction along with the terrain effect so you can see their difference easier rather than seeing them directly in the corrected gravities.

Also, take into account that the simple Bouguer correction doesn't depend on observation height because it assumes an infinite plate. This is why simple Bouguer corrections should be taken carefully, while a proper terrain correction is more suitable for most cases.

**Answer to point 2**

Your kernel is probably crashing due to lack of memory when trying to fit the deep sources. You are working with ~200000 data points and this is no small number for applying equivalent sources.
An alternative would be to use less deep sources: since they are going to be very deep, we don't need too many of them. You can do it by passing a value to the `block_size` argument of the `EquivalentSources`. By doing so your region will be divided in blocks of equal size and it will locate a single equivalent source beneath each block. This way you'll end up with less sources than before (when you put one source beneath each data point). Check the [`EquivalentSources` documentation](https://www.fatiando.org/harmonica/latest/api/generated/harmonica.EquivalentSources.html#harmonica.EquivalentSources) for a complete explanation of this behaviour or the page on [Block-averaged equivalent sources](https://www.fatiando.org/harmonica/latest/user_guide/equivalent_sources/block-averaged-eqs.html) in Harmonica's User Guide.

A second alternative would be to use the gradient-boosted equivalent sources: https://www.fatiando.org/harmonica/latest/user_guide/equivalent_sources/gradient-boosted-eqs.html

**Answer to point 3**

I really regret not including an example on how to correctly apply the correction that that Note mentions. I remember having this example somewhere but don't remember exactly where. @leouieda I think you wrote it, do you think where it can be?

BTW, this more complex example should be included in the Harmonica User Guide! (will open an issue for it)

**Reply by prince-Ha** [reply]
Thank you for your kind advice 😊


about : Problem 1
today, I try to solve the problem 


gravity data have not a observation height because this data only offer the sea level FREE AIR anomaly data & 1' topography data (https://topex.ucsd.edu/cgi-bin/get_data.cgi) so, i used a topography data as a observation height 

![image](https://user-images.githubusercontent.com/69688086/203983816-af8b8915-fa7a-40f5-95ee-00affeca9d7b.jpeg)

I can't understand why should I put the height geometric_m in coordinates I think height geometric_m and topography geometric have different height.. 

"topography geometric only caculate the density, and multiply density by height geometric_m"
Is this the right meaning?





![image](https://user-images.githubusercontent.com/69688086/203980379-35b5c826-dafd-4e53-969f-1f78f8f779fb.jpeg)

harmonica tutorial write the 1040-2670 density in the ocean but I think that bouguer correction should have (+) in ocean so, I changed the density 1040-2670 to 2670-1040 

result 
 ![image](https://user-images.githubusercontent.com/69688086/203982268-ff0ab1ab-f5bd-40e0-8eb4-05a0137ce11f.jpeg)

simple bouguer and Complete bouguer are very similar

can i change the density...?

I'm a master's course student, so I might have many misconceptions. Please understand.

thanks

**Reply by santisoler** [reply]
Hi @prince-Ha!

> I'm a master's course student, so I might have many misconceptions. Please understand.

Sure! There's no wrong question to ask. It's better to acknowledge misconceptions that carry them around. Feel free to ask for clarifications if our explanations are not clear.

> gravity data have not a observation height because this data only offer the sea level FREE AIR anomaly data & 1' topography data (https://topex.ucsd.edu/cgi-bin/get_data.cgi) so, i used a topography data as a observation height

This is a problem within the field: there are a lot of gravity datasets that don't include observation heights. This is a huge problem because the gravity field heavily depends on the observation height. The free air anomaly is not independent on the observation height. If you measure the gravity on the Earth's surface and on a plane; and compute the free-air anomaly on both observations the two resulting values will be different. If you have the data and no information about the observation height sometimes you can make the assumption that these were taken on the Earth's surface: this is the topographic height on continents and the sea level on the ocean. But take this assumption very carefully.

The problem I noticed in you notebook is that you use `data.height_geometric_m` as the observation heights. These include:
- positive heights on the continent (topographic height)
- negative heights on the ocean (bathymetry)

If you assume that the gravity observations where taken in the ground (on the continent) and on sea level (on the ocean), then your negative observation heights are wrong: you are computing the gravity field that the oceanic masses generate **on the bottom of the sea** and **not** at sea level.

> "topography geometric only caculate the density, and multiply density by height geometric_m"
Is this the right meaning?

I didn't understand this sentence. Can you rephrase or elaborate more?

> harmonica tutorial write the 1040-2670 density in the ocean but I think that bouguer correction should have (+) in ocean so, I changed the density 1040-2670 to 2670-1040

When you remove the normal gravity from the observed gravity values you obtain the gravitational attraction of every anomalous masses. It's standard to assume that the normal upper crust is composed by rocks with a density of 2670 kg/m3. In regions where you have water now those upper crust regions are occupied by water with a density 1040 kg/m3. Since these oceanic waters have less density than the same regions in the normal Earth model, they should have **negative** density contrasts, and they are (1040 - 2670) kg/m3. So, replying to your question: no, the oceanic masses should have a negative density contrast of (1040 - 2670) kg/m3.

The reason why the density change works in your case is due to the wrong reason. As I mentioned before by putting observation heights on the bathymetry depths you are computing the gravitational effect of the oceanic masses on the bottom of the sea. By changing the sign of the density of those masses you are inverting the sign of this gravitational effect, making it the same sign as the gravitational effect of those oceanic masses has **at sea level** (the one you should use). So you're trying to fix the first problem (setting the right observation heights) by introducing a wrong density contrast for the oceanic masses.

In summary, I would recommend you to solve the issue with the observation heights and use negative density contrasts for oceanic masses.

---

If you want to learn more about gravity disturbances and gravity anomalies I recommend this short article: https://www.pinga-lab.org/pdf/use-the-disturbance.pdf

One great book to understand Bouguer anomalies is "Understanding the Bouguer anomaly" by Pasteka, Mikuska and Meurers

**Reply by prince-Ha** [reply]
Hi, santi soler : )

I change the observation height

below is my assuming setting
continent height : topography data (from altimetery data  https://topex.ucsd.edu/cgi-bin/get_data.cgi)
marine height : zero

Is it a right correction?

I'll attach the file together

thanks

**Reply by santisoler** [reply]
Hi @prince-Ha. Thanks for sharing your files. Sorry I didn't have the time to read them thoroughly, but I noticed that you are now defining the height of your observation points in the following cell:

```python
height_sea_level_m = np.where(data.elevation<0, 0, data.elevation)
height_sea_level_m
```

And using these heights (converted to geometric heights) for the forward modelling step of the terrain effect:

```python
coordinates = (data.easting_m, data.northing_m, data.height_geometric_m)
data = data.assign(terrain_effect_mgal=topography_model.prism_layer.gravity(coordinates, field="g_z"), )
```

So you're on good track now! Congratulations!

**Reply by prince-Ha** [reply]
Hi @santisoler 

I was the only graduate student in my laboratory, so there was no place to solve my curiosity. But it was nice of you to reply kindly Please help me next time when you have time 

Thank you for making a good program.

**Reply by santisoler** [reply]
Glad to hear that! This is one of the things I love about building communities around the software we develop and use!

**Reply by prince-Ha** [reply]
Hi, santisoler
1.
I have a question about the Complete Bouguer correction
I used forward modeling for the topographic
https://www.fatiando.org/harmonica/latest/user_guide/topographic_correction.html

I know that the traditional Terrian correction process is the Bullard method(Bullard A, Bullard B, Bullard C)
so, forward modeling for the topographic is the first time I've heard of this method.

 - where should I refer the thesis, book, or paper about the modeling correction method 
 - I used a harmonica code so I want to refer to your code  where can I find the reference information
 - is it a right doi? (doi: 10.5281/zenodo.3628741)

2. I want to write the SCI thesis but I have no experience...  how to prepare it

happy new year thank you

**Reply by santisoler** [reply]
Hi @prince-Ha!

[Blakely (1995, pages 144-145)](https://www.cambridge.org/core/journals/geological-magazine/article/abs/blakely-r-j-1995-potential-theory-in-gravity-magnetic-applications-xix-441-pp-cambridge-new-york-port-chester-melbourne-sydney-cambridge-university-press-price-4000-us-5995-hard-covers-isbn-0-521-41508-x/8E14D29C162F311123AE27CC7ED11B96) describes the complete Bouguer correction as approximating the topography with a digital elevation model and forward modelling its gravitational effect on the observation points.

> is it a right doi? (doi: 10.5281/zenodo.3628741)

Yes, if you want to cite Harmonica in a scientific article, thesis, etc. that DOI is the one. Check out our [Citing Harmonica](https://www.fatiando.org/harmonica/latest/citing.html) page in the documentation for more details.

Happy new year for you too!

**Comment by kobebryantlin0** [comment]
您好，我是林恩伟，你发的邮件我已经收到。
